File: tests/python/test_nhom_agent_workflow.py
"""Gom nhóm cho AGENT và WORKFLOW, đúng cách skill đã làm từ lâu.

    python tests/run.py nhom_agent_workflow      (KHÔNG mạng)

Vì sao có việc này: brain dùng vài tháng là có vài chục agent và workflow, mà hai trang đó
chỉ có một danh sách phẳng - tìm bằng mắt. Skill đã có field `group` và cột nhóm; ở đây mở
đúng cơ chế ấy cho hai loại còn lại.

Ba điểm đáng canh, vì cả ba đều hỏng ÂM THẦM:

1. MỘT field cho cả ba loại. Nếu agent gọi là `category` còn skill gọi là `group` thì trang
   Studio, chỉ mục `Javis/index.md` và AI lúc tự tạo năng lực sẽ hiểu khác nhau về cùng một
   thứ, và không có lỗi nào bật lên.

2. File CŨ (viết trước bản này) không có `group`. Nó phải rơi về "Chung" chứ không được biến
   mất khỏi danh sách - người dùng chỉ phát hiện khi agent của họ không còn thấy đâu.

3. Client CŨ không gửi trường `group` lúc lưu. Server phải tự điền "Chung", đừng để
   frontmatter mọc ra `group: None`.
"""
from _paths import ROOT, SERVER  # noqa: E402,F401  - nạp server/ vào sys.path
import os
import sys
import tempfile
from pathlib import Path

os.environ.setdefault("JAVIS_STATE_DIR", tempfile.mkdtemp(prefix="javis-nhom-state-"))

import main  # noqa: E402
from starlette.testclient import TestClient  # noqa: E402

_fails = []


def check(ten, dieu_kien, them=""):
    print(("ok   " if dieu_kien else "FAIL ") + ten
          + (("  [" + str(them) + "]") if them and not dieu_kien else ""))
    if not dieu_kien:
        _fails.append(ten)


main.cfgmod.gate_active = lambda: False   # test không đụng auth thật
# base_url phải là IP: chưa bật cổng đăng nhập thì web_security siết Host để chống
# DNS-rebinding, mà "testserver" mặc định của TestClient không nằm trong allowlist.
client = TestClient(main.app, base_url="http://127.0.0.1")
BRAIN = tempfile.mkdtemp(prefix="javis-nhom-brain-")


def fm(rel):
    """Frontmatter của một file .md trong brain test."""
    return main._read_md(Path(BRAIN) / rel)[0]


def luu(loai, **form):
    """POST /agents hoặc /workflows, trả (phản hồi, frontmatter đã ghi). Slug lấy từ phản hồi
    chứ không tự đoán: _slugify của agent/workflow GIỮ dấu tiếng Việt."""
    r = client.post("/" + loai, data=dict(form, brain=BRAIN))
    slug = (r.json() or {}).get("slug", "")
    return r, fm(f"{loai}/{slug}.md"), slug


# ============================================================
# 1. Lưu qua form: nhóm vào frontmatter, thiếu nhóm thì về "Chung"
# ============================================================
r, meta, slug_ag = luu("agents", name="Viet email", role="Soạn email chăm khách", group="Marketing")
check("POST /agents nhận nhóm", r.status_code == 200 and r.json().get("ok"), r.text[:120])
check("agent: nhóm ghi vào frontmatter", meta.get("group") == "Marketing", meta)

_, meta, _ = luu("agents", name="Khong nhom", role="x")
check("agent: client cũ không gửi group → 'Chung' (không phải None)",
      meta.get("group") == main.NHOM_MAC_DINH, meta)

_, meta, _ = luu("agents", name="Nhom thua dau cach", role="x", group="  Bán hàng  ")
check("agent: nhóm được cắt khoảng trắng thừa (khỏi đẻ 'Bán hàng ' song song)",
      meta.get("group") == "Bán hàng", meta)

_, meta, _ = luu("agents", name="Nhom rong", role="x", group="   ")
check("agent: nhóm toàn khoảng trắng → 'Chung'", meta.get("group") == main.NHOM_MAC_DINH, meta)

r, meta, slug_wf = luu("workflows", name="Nghien cuu roi viet", description="hai bước",
                       steps="[]", group="Nội dung")
check("POST /workflows nhận nhóm", r.status_code == 200 and r.json().get("ok"), r.text[:120])
check("workflow: nhóm ghi vào frontmatter", meta.get("group") == "Nội dung", meta)

_, meta, _ = luu("workflows", name="Chuoi cu", steps="[]")
check("workflow: client cũ không gửi group → 'Chung'",
      meta.get("group") == main.NHOM_MAC_DINH, meta)

# ============================================================
# 2. Đọc danh sách: file CŨ chưa khai nhóm vẫn phải còn trong danh sách
# ============================================================
main._write_md(Path(BRAIN) / "agents" / "agent-cu.md",
               {"type": "agent", "name": "Agent đời cũ", "role": "viết"}, "prompt cũ")
main._write_md(Path(BRAIN) / "workflows" / "wf-cu.md",
               {"type": "workflow", "name": "Chuỗi đời cũ", "status": "off", "steps": []}, "")

ags = {a["slug"]: a for a in main.agents_index(BRAIN)}
wfs = {w["slug"]: w for w in main.workflows_index(BRAIN)}
check("agents_index trả về nhóm", ags[slug_ag]["group"] == "Marketing")
check("agents_index: file cũ chưa có group vẫn hiện, rơi vào 'Chung'",
      ags.get("agent-cu", {}).get("group") == main.NHOM_MAC_DINH)
check("workflows_index trả về nhóm", wfs[slug_wf]["group"] == "Nội dung")
check("workflows_index: file cũ chưa có group vẫn hiện, rơi vào 'Chung'",
      wfs.get("wf-cu", {}).get("group") == main.NHOM_MAC_DINH)

r = client.get(f"/agents?brain={BRAIN}")
check("GET /agents trả kèm group (trang Studio dựng cột nhóm từ đây)",
      all("group" in a for a in r.json()["agents"]))
r = client.get(f"/workflows?brain={BRAIN}")
check("GET /workflows trả kèm group",
      all("group" in w for w in r.json()["workflows"]))

# Sửa một agent rồi lưu lại: nhóm cũ phải giữ nguyên nếu form gửi lên đúng nhóm đó.
_, meta, _ = luu("agents", name="Viet email", role="Soạn email chăm khách",
                 group="Marketing", slug=slug_ag)
check("agent: lưu lại lần hai không làm mất nhóm", meta.get("group") == "Marketing", meta)

# ============================================================
# 3. Chỉ mục Javis/index.md: xếp theo nhóm cho CẢ BA loại
# ============================================================
caps = main._gather_capabilities(BRAIN)
check("caps: agent mang theo nhóm", all("group" in a for a in caps["agents"]))
check("caps: workflow mang theo nhóm", all("group" in w for w in caps["workflows"]))

idx = main._render_javis_index(caps)
check("index: có tiêu đề nhóm của agent", "### Marketing" in idx, idx[:400])
check("index: có tiêu đề nhóm của workflow", "### Nội dung" in idx)
check("index: agent nhóm Chung vẫn được liệt kê", "agent-cu" in idx)

# Brain chỉ có MỘT nhóm thì đừng thêm một tầng tiêu đề vô nghĩa.
mot_nhom = {"agents": [{"slug": "a", "name": "A", "role": "r", "model": "", "skills": [],
                        "group": main.NHOM_MAC_DINH}],
            "skills": [], "workflows": [], "loops": [], "plugins": []}
idx1 = main._render_javis_index(mot_nhom)
check("index: chỉ một nhóm thì không chèn tiêu đề nhóm thừa", "### Chung" not in idx1)

# ============================================================
# 4. Dây nối UI: BA trang dùng CHUNG một khung nhóm
# ============================================================
# Chép khung ra ba bản là ba bản trôi lệch nhau ngay lần sửa đầu (bài học của khối chọn
# skill trong màn sửa Agent - xem tests/js/test_chon_skill_va_phan_trang.js).
sj = (ROOT / "dashboard" / "studio.js").read_text(encoding="utf-8")
check("UI: có khung nhóm dùng chung", "function khungNhomHtml(" in sj and "function locTheoNhom(" in sj)
for trang, oId in (("workflows", "wfSearch"), ("agents", "agSearch"), ("skills", "skSearch")):
    check(f"UI: trang {trang} dùng khung nhóm chung", f'searchId: "{oId}"' in sj)
check("UI: form Agent gửi kèm nhóm", 'box.querySelector("#agGroup")' in sj)
check("UI: form Workflow gửi kèm nhóm", 'box.querySelector("#wfGroup")' in sj)

if _fails:
    print(f"\nFAIL {len(_fails)} muc: " + ", ".join(_fails))
    sys.exit(1)
print("\nOK - test_nhom_agent_workflow: tat ca pass")

File: tests/js/test_nhom_agent_workflow.js
/* Cột nhóm cho trang Agents và Workflows, dùng CHUNG khung với trang Skills.

       node tests/js/test_nhom_agent_workflow.js

   Trang Skills có cột nhóm từ lâu; Agents và Workflows thì không, nên brain dùng vài tháng
   là hai danh sách phẳng vài chục dòng phải dò bằng mắt. Việc này mở cột nhóm cho cả hai.

   Điều đáng canh KHÔNG phải là "có cột nhóm hay chưa" mà là CÓ ĐÚNG MỘT khung. Chép khung
   lọc ra ba bản cho ba trang thì ba bản trôi lệch nhau ngay lần sửa đầu tiên: sửa cách bỏ
   dấu ở trang này, quên hai trang kia, và người dùng gặp một trang tìm được "viet email"
   còn trang bên cạnh thì không. Đây đúng là bài học của khối chọn skill trong màn sửa Agent
   (xem test_chon_skill_va_phan_trang.js), nên phần lớn test này canh chuyện dùng chung.

   Từ 0.62.0, trình sửa TRỢ LÝ không còn ô nhóm: gom nhóm trợ lý chỉ còn một chỗ duy nhất là
   thanh nhóm ở cột trái trang Cộng sự. Phần C canh đúng điều đó, cả hai phía (form không gửi
   `group`, và trang Cộng sự bỏ miếng vá đồng bộ ngược).

   Phần cuối canh một bẫy MẤT CHỮ trong form sửa Workflow: render() chạy lại mỗi lần thêm
   hoặc xoá bước, nên ô nào lấy value từ `w` sẽ bị vẽ đè về giá trị cũ, im lặng. */
const fs = require("fs");
const path = require("path");

const ROOT = path.join(__dirname, "..", "..");
const SRC = fs.readFileSync(path.join(ROOT, "dashboard", "studio.js"), "utf8");
const WS = fs.readFileSync(path.join(ROOT, "dashboard", "workspace.js"), "utf8");

const fails = [];
const check = (name, cond, extra) => {
  console.log((cond ? "ok   " : "FAIL ") + name + (cond || extra === undefined ? "" : "  [" + extra + "]"));
  if (!cond) fails.push(name);
};

// ============================================================
// A. Chạy THẬT khung nhóm dùng chung (không chỉ soi chữ trong file)
// ============================================================
// Bóc đúng khối dùng chung rồi chạy nó với vài hàm giả (t/ic/LOC), để test bắt được lỗi
// hành vi chứ không chỉ lỗi thiếu chữ.
const i0 = SRC.indexOf("  const NHOM_MD =");
// Mốc cuối khối là tiêu đề phần Workflows. Trước đây lấy `function switchTab(`, nhưng hàm đó
// đã bỏ cùng lúc Studio ba-tab tan thành các trang riêng của rail, và mốc biến mất thì cả khối
// bóc ra rỗng - test đỏ vì lý do chẳng liên quan gì tới khung nhóm.
const i1 = SRC.indexOf("  // ===== Workflows =====");
check("tìm thấy khối khung nhóm dùng chung", i0 !== -1 && i1 > i0);

const esc = (s) => (s || "").replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#39;");
const KHOI = SRC.slice(i0, i1);
const M = new Function("esc", "t", "ic", "LOC", KHOI +
  "\n return { NHOM_MD, nhomCua, demNhom, locTheoNhom, khungNhomHtml, nhomDatalist };")(
  esc, (k) => k, () => "", () => "vi-VN");

check("nhóm mặc định là Chung", M.NHOM_MD === "Chung");
check("thiếu group → Chung", M.nhomCua({ name: "a" }) === "Chung");
check("group rỗng/toàn khoảng trắng → Chung", M.nhomCua({ group: "   " }) === "Chung");
check("group có khoảng trắng thừa được cắt", M.nhomCua({ group: " Marketing " }) === "Marketing");

const DS = [
  { slug: "viet-email", name: "Viết email", group: "Marketing", role: "soạn email" },
  { slug: "chot-don", name: "Chốt đơn", group: "Bán hàng", role: "trả lời khách" },
  { slug: "cu", name: "Agent cũ", role: "chưa xếp nhóm" },
];
const dem = M.demNhom(DS);
check("đếm theo nhóm đúng", dem["Marketing"] === 1 && dem["Bán hàng"] === 1 && dem["Chung"] === 1, JSON.stringify(dem));

const blob = (a) => `${a.name} ${a.slug} ${a.role || ""}`;
check("nhóm ALL không lọc gì", M.locTheoNhom(DS, { cat: "ALL", q: "" }, blob).length === 3);
check("lọc đúng nhóm", M.locTheoNhom(DS, { cat: "Bán hàng", q: "" }, blob).map(x => x.slug).join() === "chot-don");
check("mục chưa xếp nhóm nằm trong Chung (không biến mất)",
  M.locTheoNhom(DS, { cat: "Chung", q: "" }, blob).map(x => x.slug).join() === "cu");
// Gõ không dấu vẫn ra: người dùng gõ nhanh trên điện thoại hiếm khi bỏ dấu đầy đủ.
check("tìm không dấu ra kết quả có dấu", M.locTheoNhom(DS, { cat: "ALL", q: "viet email" }, blob).length === 1);
check("tìm hoạt động cùng lúc với lọc nhóm",
  M.locTheoNhom(DS, { cat: "Marketing", q: "chot" }, blob).length === 0);

const html = M.khungNhomHtml(DS, { cat: "ALL", q: "" },
  { bodyId: "agCards", bodyCls: "cards", searchId: "agSearch", searchPh: "tìm" });
check("khung có đủ chip nhóm", html.includes("Marketing") && html.includes("Bán hàng") && html.includes("Chung"));
check("khung có ô tìm đúng id", html.includes('id="agSearch"'));
check("khung chừa chỗ cho danh sách của từng trang", html.includes('id="agCards"') && html.includes('class="cards"'));
check("chip ALL đếm đúng tổng số", /data-cat="ALL"[\s\S]*?<span class="n">3</, html.match(/data-cat="ALL"[\s\S]{0,120}<span class="n">3</) !== null);

// Tên nhóm do người dùng gõ vào frontmatter, nên nó là chữ NGƯỜI LẠ có thể gửi qua gói .zip
// chia sẻ năng lực. Nhét thẳng vào innerHTML là một lỗ XSS.
const hiem = M.khungNhomHtml([{ slug: "x", name: "x", group: '<img src=x onerror=alert(1)>' }],
  { cat: "ALL", q: "" }, { bodyId: "b", searchId: "s", searchPh: "" });
check("tên nhóm được escape trước khi vào HTML", !hiem.includes("<img src=x"));
check("gợi ý nhóm trong form cũng escape", !M.nhomDatalist([{ group: '"><script>' }], "l").includes("<script>"));

// ============================================================
// B. CẢ BA trang dùng đúng MỘT khung đó
// ============================================================
check("khung nhóm chỉ có MỘT bản", (SRC.match(/function khungNhomHtml\(/g) || []).length === 1);
for (const [trang, state, oId] of [["Workflows", "_wfState", "wfSearch"],
                                   ["Agents", "_agState", "agSearch"],
                                   ["Skills", "_skState", "skSearch"]]) {
  check(`trang ${trang} có trạng thái nhóm + ô tìm riêng`,
    SRC.includes(`const ${state} = { cat: "ALL", q: ""`) && SRC.includes(`searchId: "${oId}"`));
  check(`trang ${trang} vẽ bằng khung chung`,
    new RegExp(`khungNhomHtml\\(all, ${state},`).test(SRC));
  check(`trang ${trang} nối sự kiện bằng hàm chung`,
    new RegExp(`ganKhungNhom\\(panel, ${state},`).test(SRC));
}
// Không còn khung riêng của trang Skills (lớp .sk2 cũ) sống song song với khung chung.
check("CANARY: không còn khung nhóm riêng của trang Skills", !SRC.includes('class="sk2-side"'));

check("thẻ agent hiện nhóm của nó", SRC.includes('class="ag-group"'));
check("hàng workflow hiện nhóm của nó", SRC.includes('class="wf-group"'));
check("bấm Chọn tất cả chỉ lấy mục ĐANG HIỆN (đúng nhóm + đúng ô tìm)",
  SRC.includes("_wfFiltered().map(w => w.slug)") && SRC.includes("_agFiltered().map(a => a.slug)"));

// ============================================================
// C. Form: chọn nhóm được, và gửi kèm lúc lưu
// ============================================================
for (const [ten, oId, list] of [["Workflow", "wfGroup", "wfGroupList"]]) {
  check(`form ${ten} có ô nhập nhóm`, SRC.includes(`id="${oId}"`));
  check(`form ${ten} gợi ý nhóm đang có (khỏi đẻ Marketing và marketing song song)`,
    SRC.includes(`nhomDatalist(`) && SRC.includes(`"${list}"`));
}
check("lưu Workflow gửi kèm group", /group: nhom\.trim\(\) \|\| NHOM_MD/.test(SRC));

// ---- Trình sửa TRỢ LÝ không còn ô nhóm (0.62.0) ----
// Chủ repo 21/09: "xoá nhóm ở đây vì đã có phần gom nhóm rồi". Hai chỗ cùng đặt một field thì
// chỗ nào cũng ghi đè được chỗ kia: đổi nhóm ở cột trái rồi bấm Lưu trong form là nhóm nhảy
// về giá trị cũ, im lặng - đúng lỗi phải vá bằng dongBoNhomForm() suốt từ 0.59.2. Gom nhóm
// nay chỉ còn MỘT chỗ, nên miếng vá đó cũng phải biến mất theo.
check("form Agent KHÔNG còn ô nhóm",
  !SRC.includes('id="agGroupSel"') && !SRC.includes('id="agGroup"'));
check("lưu Agent KHÔNG gửi group (server giữ nguyên nhóm đang có)",
  !/group: nhomLuu\(/.test(SRC));
check("trang Cộng sự bỏ luôn miếng vá đồng bộ nhóm vào form",
  !WS.includes("function dongBoNhomForm"));
check("gom nhóm trợ lý vẫn còn NGUYÊN ở cột trái (thanh nhóm + menu Chuyển sang nhóm)",
  WS.includes('id="wsGroup"') && /nutMenu\("folder", t\("ws\.move_group"\)/.test(WS));

// ---- Bẫy mất chữ trong form Workflow ----
// render() chạy lại mỗi lần thêm/xoá/đảo bước. Ô nào đọc value từ `w` sẽ bị vẽ đè về giá trị
// cũ: gõ tên workflow rồi bấm "+ Bước" là mất tên, không một thông báo nào.
const iw = SRC.indexOf("async function editWorkflow(");
const FNW = SRC.slice(iw, SRC.indexOf("\n  // ===== Agents =====", iw));
check("form Workflow giữ tên/mô tả/nhóm trong biến, không đọc lại từ w mỗi lần vẽ",
  /let ten = w \? \(w\.name \|\| ""\) : "";/.test(FNW) && /let nhom = w \? nhomCua\(w\) : NHOM_MD;/.test(FNW));
check("captureSteps() hứng cả ba ô trước khi vẽ lại",
  /if \(oNe\) ten = oNe\.value;/.test(FNW) && /if \(oNh\) nhom = oNh\.value;/.test(FNW));
check("CANARY: ô tên/mô tả không còn lấy giá trị thẳng từ w",
  !FNW.includes('value="${esc(w ? w.name : "")}"') && !FNW.includes('value="${esc(w ? w.description : "")}"'));

if (fails.length) {
  console.log(`\nFAIL ${fails.length} muc: ` + fails.join(", "));
  process.exit(1);
}
console.log("\nOK - test_nhom_agent_workflow: tat ca pass");
